lib/asciidoctor/reader.rb

```diff
@@ -837,10 +837,42 @@ def preprocess_include_directive raw_target, raw_attributes
         warn %(asciidoctor: ERROR: #{line_info}: maximum include depth of #{@maxdepth[:rel]} exceeded)
         return
       elsif ::RUBY_ENGINE_OPAL && ::JAVASCRIPT_IO_MODULE == 'xmlhttprequest'
-        # NOTE resolves uri relative to currently loaded document
-        # NOTE we defer checking if file exists and catch the 404 error if it does not
-        target_type = :file
-        inc_path = relpath = @include_stack.empty? && ::Dir.pwd == @document.base_dir ? target : %(#{@dir}/#{target})
+        # NOTE the only way to check if the file exists is to catch a 404 response in the IO module
+        target, target_type, base_dir = ((@path_resolver ||= PathResolver.new '\\').posixify target), :file, @document.base_dir
+        if target.start_with? 'file://'
+          inc_path = relpath = target
+        elsif Helpers.uriish? target
+          # NOTE only require allow-uri-read if cross origin
+          unless (@path_resolver.descends_from? target, base_dir) || (@document.attributes.key? 'allow-uri-read')
+            replace_next_line %(link:#{target}[])
+            return true
+          end
+          inc_path = relpath = target
+        elsif @path_resolver.absolute_path? target
+          inc_path = relpath = ((target.start_with? '/') ? 'file://' : 'file:///') + target
+        elsif (ctx_dir = (top_level = @include_stack.empty?) ? base_dir : @dir) == '.'
+          # NOTE won't work in the Firefox web extension (unless we use the second line)
+          inc_path = relpath = target
+          #inc_path = (File.dirname `window.location.href`) + '/' + (relpath = target)
+        elsif (ctx_dir.start_with? 'file://') || !(Helpers.uriish? ctx_dir)
+          # WARNING a relative ctx_dir won't work in the Firefox web extension
+          inc_path = ctx_dir + '/' + target
```

Comment:
`ctx_dir` can be relative in the Firefox web extension ? could you please elaborate ?

Reply:
It won't work unless you set base_dir to an absolute URI for the reason in the comment a few lines above. If we didn't want that requirement, we'd have to expand `.` to the dirname of window.location.href.

Reply:
I see... but I think Asciidoctor.js is best suited to expand/resolve a relative path in a browser environment

Reply:
I agree, that's what we're going to depend on here. I'm just noting that here so I remember what the assumption is.